### PR TITLE
Visual indication of outdated VCS snapshot

### PR DESCRIPTION
**Pull request by galin-enso**
### Pull Request Description

This PR provides a visual indication of whether the project's current state differs from the most recent snapshot saved in the VCS. The project name displayed in the IDE changes to a darker text to indicate that the VCS snapshot is outdated, and back to a lighter text when the current project state corresponds to the last saved VCS snapshot.

https://user-images.githubusercontent.com/117099775/208088438-20dfc2aa-2a7d-47bf-bc12-3d3dff7a4974.mp4

The outdated project snapshot indicator is set when:
* A node is moved.
* A node is added or removed.
* The text editor is used to edit the text.
* The project is auto-saved, and the auto-saved project state does not correspond to the last saved snapshot in the VCS.

The outdated project snapshot indicator is cleared when:
* A new project snapshot is successfully saved using `ctrl+s`.
* The project is auto-saved, and the auto-saved project state is confirmed to correspond to the last saved snapshot in the VCS. This occurs, for example, when a project change is undone and the project is reverted to the last saved snapshot state.

The auto-save events do not occur immediately after a project change but have a short delay, thus the VCS status update is affected by the same delay when triggered by an auto-save event.

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -74,6 +74,8 @@
 - [Added scroll bounce animation][3836] which activates when scrolling past the
   end of scrollable content.
 - [Added project snapshot saving on shortcut][3923]
+- [Added a visual indication whether the last saved project snapshot is up to
```
Comment (wdanilo) [inline CHANGELOG.md]: This should be explained better. This is a user-facing description - please make it more wordy un such a way that someone outside of our team would understand it.

File: app/gui/src/controller/project.rs [hunk 2]
```diff
@@ -260,6 +260,29 @@ impl Project {
             Ok(())
         }
     }
+
+    /// Check whether the current state of the project differs from the last snapshot in the VCS.
+    #[profile(Detail)]
+    pub fn check_project_vcs_is_outdated(&self) -> impl Future<Output = FallibleResult<bool>> {
```
Comment (wdanilo) [inline app/gui/src/controller/project.rs]: I know this is already a long name, but it is not a correct english phrase. Maybe we sohuld renaem it to `chech_if_project_vcs_is_outdated` ?

File: app/gui/src/controller/project.rs [hunk 3]
```diff
@@ -260,6 +260,29 @@ impl Project {
...
+        let project_root_id = self.model.project_content_root_id();
+        let path_segments: [&str; 0] = [];
```
Comment (wdanilo) [inline app/gui/src/controller/project.rs]: can we use default here?

File: app/gui/src/controller/project.rs [hunk 4]
```diff
@@ -260,6 +260,29 @@ impl Project {
...
+        let root_path = Path::new(project_root_id, &path_segments);
+        let language_server = self.model.json_rpc();
+        async move {
+            let status = language_server.vcs_status(&root_path).await;
+            if let Err(RpcError::RemoteError(json_rpc::messages::Error {
+                code: error_code, ..
+            })) = status
+            {
+                if error_code == code::FILE_NOT_FOUND {
+                    // No VCS found, thus current project state is saved in a snapshot.
```
Comment (wdanilo) [inline app/gui/src/controller/project.rs]: I dont understand how this code is connected to this comment. I mean, I see the error code, but why `Ok(true)` means "project state is saved in a snapshot?

File: app/gui/src/controller/project.rs [hunk 5]
```diff
@@ -338,11 +362,11 @@ mod tests {
...
             message:   "message".into(),
         };
-        let vcs_err = RpcError::RemoteError(json_rpc::messages::Error {
-            code:    code::FILE_NOT_FOUND,
-            message: "VCS is not initialized.".into(),
-            data:    None,
-        });
+        let vcs_status = language_server::response::VcsStatus {
+            dirty:     false,
+            changed:   Vec::new(),
```
Comment (wdanilo) [inline app/gui/src/controller/project.rs]: these can be defaulted

File: app/gui/src/controller/project.rs [hunk 6]
```diff
@@ -356,28 +380,63 @@ mod tests {
...
+                Ok(vcs_status_clone)
+            } else {
+                Err(vcs_err)
+            }
+        });
+
+        // Expect a 2nd call to `vcs_status` after a `write_vcs` is performed.
+        json_client.expect.vcs_status(move |path| {
+            assert_eq!(path, &root_path);
+            assert!(vcs.init.get(), "VCS is not initialized.");
```
Comment (wdanilo) [inline app/gui/src/controller/project.rs]: I see this string 3rd time. Please extract such things to constants

File: app/gui/src/presenter/project.rs [hunk 7]
```diff
@@ -171,9 +171,33 @@ impl Model {
...
+            }
+        })
+    }
+
+    /// This is called on any module change. It should be followed by a call to
+    /// `check_project_changed` to verify that there is indeed a project change compared to the last
+    /// saved VCS snapshot.
+    fn assume_project_changed(&self) {
+        self.view.graph().model.breadcrumbs.set_project_changed(true);
+    }
```
Comment (wdanilo) [inline app/gui/src/presenter/project.rs]: 1. If it sohuld be followed by a call to something, why we dont create one function that does both?
2. The only occurence of this function is not followed by `check_project_changed`! At least I cant see it.
3. I dont understand the docs - they tell that we should call `check_project_changed` to "verify there is indeed a project change". Why is that? Why this function even exists if we need an additional verification? This is IMO strange logic here.

**Comment by wdanilo** [comment]
Also, lets make the visual indicator more subtle. Instead of the yellow color, lets make it more transparent when the project is saved and less trnsparent (more black) when it is not.

File: app/gui/src/controller/project.rs (hunk 4 above)
Comment (farmaazon) [reply app/gui/src/controller/project.rs]: I wonder is we should not initialize the VCS on project start, so we always have some commit we can revert to. In that scenario, this branch should print error, not only return true.

File: app/gui/src/controller/project.rs [hunk 8]
```diff
@@ -260,6 +260,29 @@ impl Project {
...
+                code: error_code, ..
+            })) = status
+            {
+                if error_code == code::FILE_NOT_FOUND {
+                    // No VCS found, thus current project state is saved in a snapshot.
+                    return Ok(true);
+                }
+            }
+            let status = status?;
+            Ok(status.dirty)
```
Comment (farmaazon) [inline app/gui/src/controller/project.rs]: I wonder if something like that would work:
```suggestion
            if let Err(RpcError::RemoteError(json_rpc::messages::Error {
                code: code::FILE_NOT_FOUND, ..
            })) = status
            {
                // No VCS found, thus current project state is saved in a snapshot.
                Ok(true);
            } else {
                Ok(status?.dirty)
            }
```

File: app/gui/src/controller/project.rs (hunk 6 above)
Comment (farmaazon) [reply app/gui/src/controller/project.rs]: > I see this string 3rd time. Please extract such things to constants

Please do it in the error responses only. Let's not obfuscate the test assertions.

File: app/gui/src/model/project.rs [hunk 9]
```diff
@@ -168,6 +168,8 @@ pub type Synchronized = synchronized::Project;
 pub enum Notification {
     /// One of the backend connections has been lost.
     ConnectionLost(BackendConnection),
+    /// A successful auto-save action is performed.
+    TextAutoSave,
```
Comment (farmaazon) [inline app/gui/src/model/project.rs]: Why is this `TextAutoSave`, not just `AutoSave`?

If there is a reason, I don't see it in the doc.

File: app/gui/src/presenter/project.rs [hunk 10]
```diff
@@ -303,14 +327,30 @@ impl Project {
...
-                    crate::BACKEND_DISCONNECTED_MESSAGE,
+            match notification {
+                model::project::Notification::ConnectionLost(_) => {
+                    let message = crate::BACKEND_DISCONNECTED_MESSAGE;
+                    let message = view::status_bar::event::Label::from(message);
+                    model.status_bar.add_event(message);
+                }
+                model::project::Notification::TextAutoSave => {
+                    model.check_project_changed();
+                }
```
Comment (farmaazon) [inline app/gui/src/presenter/project.rs]: I would refactor it a bit. I think the logic of "check VCS status after autosave" should be covered in models/controllers, not in presenter. I prefer models: there could be additional notification (next to TextAutoSaved) about VCSStatus Change which is sent on each module change + the model::Project is the one calling check_vcs_status and sending a notification when it returns.

There should be a notification about "VCS status change", and here we just read it.

**Review by farmaazon — CHANGES_REQUESTED** [review]

File: app/gui/src/controller/project.rs (hunk 4 above)
Comment (galin-enso) [reply app/gui/src/controller/project.rs]: Oh, this is a typo in the comment! It should be `No VCS found, thus current project state is NOT saved in a snapshot.` and it returns the equivalent of a VCS being available but with an outdated snapshot.

Comment (galin-enso) [reply app/gui/src/controller/project.rs]: @farmaazon That makes sense, I'll change it.

File: app/gui/src/model/project/synchronized.rs [hunk 11]
```diff
@@ -167,6 +169,44 @@ impl ContentRoots {
...
+            publisher.notify(model::project::Notification::VcsStatusChanged(
+                model::project::VcsStatus::Dirty,
+            ));
+        }
+        Ok(language_server::response::VcsStatus { dirty: false, .. }) => {
+            publisher.notify(model::project::Notification::VcsStatusChanged(
+                model::project::VcsStatus::Clean,
+            ));
+        }
+    }
```
Comment (farmaazon) [inline app/gui/src/model/project/synchronized.rs]: We can deduplicate things here somehow by extracting notify call and notification constructing:
```rust
let status = match &status {...}
publisher.notify(model::project::Notification::VcsStatusChanged(status);
```

**Review by farmaazon — APPROVED** [review]

File: app/gui/src/model/project/synchronized.rs (hunk 11 above)
Comment (galin-enso) [reply app/gui/src/model/project/synchronized.rs]: Good point, I didn't see it. Fixed now.

**Review by vitvakatu — APPROVED** [review]
QA passed

**Pull request merged**